### ISSUE TITLE
bulk requests to openCPU

### ISSUE THREAD

**Issue by rubenarslan** [rubenarslan/formr.org#137]
except (maybe) dynamic values, it would make sense to send commands (labels and code execution) to openCPU in a bundle, so we avoid the significant overhead that occurs when things are written to disk there.

Several ways to do it, here for the example of labels:

1. Separate the commands completely, and then later we access the results via HTTP. Seems clean, but still quite a lot of HTTP overhead and we potentially send the same data twice.
```r
variable1_label = formr_render("`r survey$x`")
variable2_label = formr_render("`r survey$y`")
```

2. Implode all labels into one big command with some unique splitter like `~~~formr_splitter~~~`, run the dataNeeded stuff on that, send it in one go and split it up on the PHP side, using the order in the result to see which item/choice it belongs to.
Probably cheapest in terms of overhead, but a little unclean. 

I favour 2, I think.

Error handling will become a mess for all of these ways though. Not sure how to deal with that.

**Comment by cyriltata** [comment]
This is how I think it should be:

From the php side a box is created with the strings to parse something like so
<pre>
$strings_to_parse = array(
     'opencpu_key1' => array(
          'key' => 'opencpu_key1',
          'string' => 'string_to_parse',
          'data' => $required_data1
     ),
    'opencpu_key2' => array(
          'key' => 'opencpu_key2',
          'string' => 'string_to_parse2',
          'data' => $required_data2
     ),
);
</pre>

and we can add a helper function for this like the other opencpu helper functions
`
$parsed_strings = opencpu_bulk_parse($strings_to_parse);
`

then in the formr package the function that handles this iterates over each item and then tries to 'parse' a string using it's associated data. The opencpu_key1, opencpu_key2 are unique ids that the client expects to receive back in order to uniquely identify the parsed strings. When returning the data you can add an index to each block say _parsed_string_ to contain the parsed or if an error occurred when parsing a particular string (i.e processing an item in the block) add an index _parse_error_ . This way the client will check which were successful and which weren't using that and proceed accordingly. I am not a R expect but I think there should be something similar to try..catch in R to gracefully handle errors for each item and not cause the whole operation to crash

**Comment by rubenarslan** [comment]
That's a workable idea, agreed. 
However, we can and should reduce overhead further, by making R do the knitting only once. To demonstrate the overhead, see this:

```r
# calling rnorm is really cheap
system.time({ rnorm(1) } )
#   user  system elapsed 
#      0       0       0 
# however calling knitr adds some overhead (and this is on my SSD, it's significantly slower on hard disk)
system.time({ formr_render("`r rnorm(1)`") } )
#   user  system elapsed 
#  0.041   0.020   0.074 

# however, when we simply chain the document together with a separator (which we later explode), the overhead doesn't add up. The user would have to use Rs debugging capabilities, but I think it would be ok.
system.time({ formr_render("`r rnorm(1)`XXXformr_seperatorXXXXMD5_KEY_OF_REQUEST`r rpois(1,1)`") } )
#   user  system elapsed 
# 0.037   0.020   0.060 
```

**Comment by rubenarslan** [comment]
To flesh this out a little, we should do this:

1. take unique knitting requests (for all item and choice labels).

2. get their md5 hash

3. concatenate the results of 1 and 2 like this 
    ```
    MD5HASH1
    ~~~formr_md5_split~~~
    `r REQUEST1`
    ~~~formr_label_split~~~
    MD5HASH2
    ~~~formr_md5_split~~~
    `r REQUEST2`
    ```

4. send this to openCPU
5. get result in its entirety
6. split strings along ~~~formr_label_split~~~ and then along ~~~formr_md5_split~~~
7. get strings

we could also do away with the md5 hashing, I thought we might need this for having a good cache but your new "submit whole page" approach might obviate that

**Comment by cyriltata** [comment]
check this out https://github.com/rubenarslan/formr.org/pull/145

**Comment by rubenarslan** [comment]
doing this successfully now, yay
